File: src/main/java/com/facimus/procesos/ejecucion/service/CasoService.java
package com.facimus.procesos.ejecucion.service;

import java.util.List;
import java.util.Map;

import org.springframework.data.domain.Pageable;

import com.facimus.procesos.common.api.PageResponse;
import com.facimus.procesos.ejecucion.dto.response.CasoDetalleResponse;
import com.facimus.procesos.ejecucion.dto.response.CasoResponse;
import com.facimus.procesos.ejecucion.dto.response.EventoCasoResponse;
import com.facimus.procesos.ejecucion.model.EstadoCaso;

/**
 * Los casos: una ejecucion de una version publicada. Abrir, mirar y cerrar; quien los hace avanzar es el motor, y
 * toda operacion que los toca empieza bloqueando su fila (D3).
 */
public interface CasoService {

    /**
     * Abre un caso sobre la version vigente del proceso y lo avanza hasta donde llegue: normalmente, hasta la
     * primera tarea de la bandeja.
     */
    CasoResponse abrir(Long empresaId, Long usuarioId, Long procesoId, String referencia,
            Map<String, Object> variables);

    PageResponse<CasoResponse> listar(Long empresaId, Long procesoId, EstadoCaso estado, String referencia,
            Pageable pagina);

    /** El caso con por donde ha pasado y con que variables decide. */
    CasoDetalleResponse obtener(Long empresaId, Long casoId);

    /** La linea de tiempo del caso, en el orden en que pasaron las cosas. */
    List<EventoCasoResponse> eventos(Long empresaId, Long casoId);

    /** Cierra el caso antes de tiempo: sus tokens vivos se apagan y no vuelve a moverse (R-51). */
    CasoResponse cancelar(Long empresaId, Long usuarioId, Long casoId);

    /** Reemplaza las variables de un caso, normalmente para corregir lo que dejo a un gateway sin camino. */
    CasoResponse corregirVariables(Long empresaId, Long casoId, Map<String, Object> variables, Long version);

    /** Vuelve a evaluar lo que dejo el caso en ERROR, con las variables que tenga ahora. */
    CasoResponse reintentar(Long empresaId, Long usuarioId, Long casoId);
}

File: frontend/src/app/service/caso.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../environments/environment';
import {
  AbrirCasoRequest,
  Caso,
  CasoDetalle,
  EventoCaso,
  FiltrosCaso,
  VariablesRequest,
} from '../models/caso.model';
import { PageResponse } from '../models/page-response.model';

/** Los casos de la tienda: cada uno es una ejecucion de una version publicada, es decir un pedido. */
@Injectable({ providedIn: 'root' })
export class CasoService {
  private readonly http: HttpClient = inject(HttpClient);
  private readonly api: string = `${environment.apiUrl}/api/v1`;

  /** Los mas nuevos primero. Los filtros vacios no se envian. */
  listar(filtros: FiltrosCaso): Observable<PageResponse<Caso>> {
    const params: Record<string, string | number> = {
      pagina: filtros.pagina,
      orden: `${filtros.orden},${filtros.direccion}`,
    };
    if (filtros.procesoId !== '') {
      params['procesoId'] = filtros.procesoId;
    }
    if (filtros.estado !== '') {
      params['estado'] = filtros.estado;
    }
    if (filtros.referencia.trim()) {
      params['referencia'] = filtros.referencia.trim();
    }
    return this.http.get<PageResponse<Caso>>(`${this.api}/casos`, { params });
  }

  /** El caso con sus pasos y sus variables. */
  obtener(id: number): Observable<CasoDetalle> {
    return this.http.get<CasoDetalle>(`${this.api}/casos/${id}`);
  }

  /** La linea de tiempo: cada decision, cada tarea y cada error, en el orden en que pasaron. */
  eventos(id: number): Observable<EventoCaso[]> {
    return this.http.get<EventoCaso[]>(`${this.api}/casos/${id}/eventos`);
  }

  /**
   * Abre un caso sobre la version vigente y lo corre hasta que se detiene. El proceso tiene que empezar con un
   * evento de inicio normal: uno que empieza con un mensaje se abre mandando ese mensaje.
   */
  abrir(procesoId: number, solicitud: AbrirCasoRequest): Observable<Caso> {
    return this.http.post<Caso>(`${this.api}/procesos/${procesoId}/casos`, solicitud);
  }

  /** Lo cierra antes de tiempo: sus tokens vivos se apagan y deja de moverse. No se borra nada. */
  cancelar(id: number): Observable<Caso> {
    return this.http.post<Caso>(`${this.api}/casos/${id}/cancelar`, null);
  }

  /** Reemplaza las variables enteras, sobre la version que se leyo. */
  variables(id: number, solicitud: VariablesRequest): Observable<Caso> {
    return this.http.patch<Caso>(`${this.api}/casos/${id}/variables`, solicitud);
  }

  /** Vuelve a evaluar lo que lo dejo sin camino, con las variables que tiene ahora. */
  reintentar(id: number): Observable<Caso> {
    return this.http.post<Caso>(`${this.api}/casos/${id}/reintentar`, null);
  }
}
